### PR TITLE
feat: Add discrete diffusion LLM (dLLM) supervised fine-tuning support

### PR DESCRIPTION
**Pull request by zyzhou5**
# What does this PR do ?

Add supervised fine-tuning (SFT) support for diffusion language models (dLLMs) to Automodel. Instead of next-token prediction, dLLM trains a model to denoise corrupted tokens. Each training step corrupts a subset of supervised tokens and the model predicts the clean token at each position. 

# Changelog

 **dLLM SFT recipe and components**
  - Add `DiffusionLMSFTRecipe` extending the LLM fine-tuning recipe (`nemo_automodel/recipes/dllm/train_ft.py`)
  - Add pluggable strategy pattern (`nemo_automodel/recipes/dllm/strategy.py`) so new dLLM variants can be added without modifying the recipe
  - Add `MDLMCrossEntropyLoss` with inverse-corruption-probability weighting and unified `DLLMLossOutput` return type
  (`nemo_automodel/components/loss/dllm_loss.py`)
  - Add `corrupt_uniform` and `corrupt_blockwise` corruption functions with vectorized Gumbel-max top-k selection
  (`nemo_automodel/components/datasets/dllm/corruption.py`)
  - Add `DLLMCollator` for two-stage block-aligned padding (block fill + global pad) (`nemo_automodel/components/datasets/dllm/collate.py`)
  - Add `dllm_sft` command to CLI (`nemo_automodel/_cli/app.py`)
 
  **Dataset support**
  - Add `unshifted` format to `ChatDataset` and formatting utilities for dLLM training (returns `input_ids` + `loss_mask` instead of shifted `input_ids`/`labels`)

 **FSDP2**
  - Add `autocast_dtype` field to `FSDP2Config` for wrapping forward pass in `torch.autocast`
  - Add YAML-to-object instantiation for `mp_policy`, `offload_policy`, and `autocast_dtype` in `_dist_setup.py`
  - Skip `model.to(device)` when FSDP2 CPU offload is active (FSDP manages GPU placement during forward/backward)

  **Tests**
  - Add unit tests for `DLLMCollator` (block alignment, two-stage padding, `max_seq_len` cap)
  - Add unit tests for `corrupt_uniform` and `corrupt_blockwise` (output shapes, mask invariants, block-level sampling)
  - Add unit tests for `MDLMCrossEntropyLoss` (normalization, zero-loss invariant)

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?

If you haven't finished some of the above items you can still open "Draft" PR.

### CODE REVIEW

File: nemo_automodel/_transformers/auto_model.py [hunk 1]
```diff
@@ -766,19 +765,33 @@ class NeMoAutoModelForTextToWaveform(_BaseNeMoAutoModelClass, AutoModelForTextTo
     pass
 
 
-class _NeMoAutoModelForRetrievalBase:
-    """Private shared base for encoder auto-models.
+class NeMoAutoModelBiencoder:
```
Comment (akoumpa) [inline nemo_automodel/_transformers/auto_model.py]: mm I think you may want to rebase / merge latest main

Comment (pthombre) [reply nemo_automodel/_transformers/auto_model.py]: @zyzhou5 Please do this

File: nemo_automodel/_cli/app.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
+#!/usr/bin/env python3
```
Comment (akoumpa) [inline nemo_automodel/_cli/app.py]: I think this can be removed, we refactored the cli recently, sorry for the trouble.

File: nemo_automodel/components/datasets/llm/formatting_utils.py [hunk 3]
```diff
@@ -178,9 +179,47 @@ def _package_tokenized_example(
...
+        # No shift: input_ids stays at full length, loss_mask = assistant_masks.
+        loss_mask = [int(bool(m)) for m in assistant_masks]
+
+        # Compute content length (skip trailing pad tokens).
+        content_length = len(input_ids)
+        if pad_token_id is not None and content_length > 0:
+            end = content_length
+            while end > 0 and input_ids[end - 1] == pad_token_id:
+                end -= 1
+            if pad_token_id == eos_token_id:
```
Comment (akoumpa) [inline nemo_automodel/components/datasets/llm/formatting_utils.py:200]: nice
